sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -0,0 +1,455 @@
+  tableFormatConfig.setBlockCache(new LRUCache(conf.blockCacheSizeMB * 1024 * 1024))
+  tableFormatConfig.setFilterPolicy(bloomFilter)
+
+  private val dbOptions = new Options() // options to open the RocksDB
+  dbOptions.setCreateIfMissing(true)
+  dbOptions.setTableFormatConfig(tableFormatConfig)
+  private val dbLogger = createLogger() // for forwarding RocksDB native logs to log4j
+  dbOptions.setStatistics(new Statistics())
+
+  private val workingDir = createTempDir("workingDir")
+  private val fileManager = new RocksDBFileManager(
+    dfsRootDir, createTempDir("manager"), hadoopConf, loggingId = loggingId)
+  private val byteArrayPair = new ByteArrayPair()
+  private val commitLatencyMs = new mutable.HashMap[String, Long]()
+  private val acquireLock = new Object
+
+  @volatile private var db: NativeRocksDB = _
+  @volatile private var loadedVersion = -1L   // -1 = nothing valid is loaded
+  @volatile private var numCommittedKeys = 0L
+  @volatile private var numUncommittedKeys = 0L
+  @volatile private var acquiredThreadInfo: AcquiredThreadInfo = _
+
+  workingDir.mkdirs()
+
+  /**
+   * Load the given version of data in a native RocksDB instance.
+   * Note that this will copy all the necessary file from DFS to local disk as needed,
+   * and possibly restart the native RocksDB instance.
+   */
+  def load(version: Long): RocksDB = {
+    assert(version >= 0)
+    acquire()
+    logInfo(s"Loading $version")
+    try {
+      if (loadedVersion != version) {
+        closeDB()
+        val metadata = fileManager.loadCheckpointFromDfs(version, workingDir)
+        openDB()
+        numUncommittedKeys = metadata.numKeys
+        numCommittedKeys = metadata.numKeys
+        loadedVersion = version
+      }
+      writeBatch.clear()
+      logInfo(s"Loaded $version")
+    } catch {
+      case t: Throwable =>
+        loadedVersion = -1  // invalidate loaded data
+        throw t
+    }
+    this
+  }
+
+  /**
+   * Get the value for the given key if present, or null.
+   * @note This will return the last written value even if it was uncommitted.
+   */
+  def get(key: Array[Byte]): Array[Byte] = {
+    writeBatch.getFromBatchAndDB(db, readOptions, key)
+  }
+
+  /**
+   * Put the given value for the given key.
+   * @note This update is not committed to disk until commit() is called.
+   */
+  def put(key: Array[Byte], value: Array[Byte]): Array[Byte] = {
+    val oldValue = writeBatch.getFromBatchAndDB(db, readOptions, key)
+    writeBatch.put(key, value)
+    if (oldValue == null) {
+      numUncommittedKeys += 1
+    }
+    oldValue
+  }
+
+  /**
+   * Remove the key if present, and return the previous value if it was present (null otherwise).
+   * @note This update is not committed to disk until commit() is called.
+   */
+  def remove(key: Array[Byte]): Array[Byte] = {
+    val value = writeBatch.getFromBatchAndDB(db, readOptions, key)
+    if (value != null) {
+      writeBatch.remove(key)
+      numUncommittedKeys -= 1
```

Comment:
Same here, if the value could be from committed db, how can we always decrease `numUncommittedKeys`?

Reply:
This also works as well regardless of where it comes. If it's removed from write batch, it should either from committed RocksDB (key is going to be removed) or uncommitted write batch (numUncommittedKeys should be increased in prior).

Reply:
We should consider the `num[Unc/C]ommittedKeys` as a kind of status here, the number here means the absolute value from the beginning(version 0) to now. So that's also why there's no place we set the `numUncommittedKeys` to 0.